service/view_building_coordinator.cc

```diff
@@ -0,0 +1,200 @@
+    std::set<view_name> views_to_add;
+    for (auto& view: views) {
+        if (!_db.find_keyspace(view.first).uses_tablets() || std::find(built.begin(), built.end(), view) != built.end()) {
+            continue;
+        }
+
+        auto base_id = get_base_id(view);
+        if (!state.tasks.contains(base_id) || !state.tasks.at(base_id).contains(view)) {
+            views_to_add.insert(view);
+        }
+    }
+    return views_to_add;
+}
+
+std::set<view_name> view_building_coordinator::get_views_to_remove(const vbc_state& state, const std::vector<view_name>& views) {
+    std::set<view_name> views_to_remove;
+    for (auto& [_, view_tasks]: state.tasks) {
+        for (auto& [view, _]: view_tasks) {
+            if (std::find(views.begin(), views.end(), view) == views.end()) {
+                views_to_remove.insert(view);
+            }
+        }
+    }
+    return views_to_remove;
+}
+
+future<std::vector<canonical_mutation>> view_building_coordinator::add_view(const group0_guard& guard, const view_name& view_name) {
+    vbc_logger.info("Register new view: {}.{}", view_name.first, view_name.second);
+
+    auto base_id = get_base_id(view_name);
+    auto& base_cf = _db.find_column_family(base_id);
+    auto erm = base_cf.get_effective_replication_map();
+    auto& tablet_map = erm->get_token_metadata().tablets().get_tablet_map(base_id);
+
+    std::vector<canonical_mutation> muts;
+    for (auto tid = std::optional(tablet_map.first_tablet()); tid; tid = tablet_map.next_tablet(*tid)) {
+        const auto& tablet_info = tablet_map.get_tablet_info(*tid);
+        auto range = tablet_map.get_token_range(*tid);
+
+        for (auto& replica: tablet_info.replicas) {
+            auto mut = co_await _sys_ks.make_vbc_task_mutation(guard.write_timestamp(), view_name, replica.host, replica.shard, range);
+            muts.emplace_back(std::move((mut)));
+        }
+    }
+    co_return muts;
+}
+
+future<std::vector<canonical_mutation>> view_building_coordinator::remove_view(const group0_guard& guard, const view_name& view_name) {
+    vbc_logger.info("Unregister all remaining tasks for view: {}.{}", view_name.first, view_name.second);
+    
+    auto muts = co_await _sys_ks.make_vbc_remove_view_tasks_mutations(guard.write_timestamp(), view_name);
+    co_return std::vector<canonical_mutation>{muts.begin(), muts.end()};
+}
+
+future<> run_view_building_coordinator(std::unique_ptr<view_building_coordinator> vb_coordinator, replica::database& db, raft_group0& group0) {
+    std::exception_ptr ex;
+    db.get_notifier().register_listener(vb_coordinator.get());
+    try {
+        co_await with_scheduling_group(group0.get_scheduling_group(), [&] {
+            return vb_coordinator->run();
+        });
+    } catch (...) {
+        ex = std::current_exception();
+    }
+    if (ex) {
+        on_fatal_internal_error(vbc_logger, format("unhandled exception in view_building_coordinator::run(): {}", ex));
```

Comment:
This is way too punishing. A bug in the view building coordinator is bad, but shouldn't take down the cluster.

We can even retry in case it was a transient failure (with appropriate backoff).